guardian-ui/src/GuardianApi.ts

```diff
@@ -0,0 +1,236 @@
+        websocketUrl,
+        requestTimeoutMs,
+        (error: JsonRpcError) => {
+          console.error('failed to create websocket', error);
+          reject(error);
+          this.shutdown();
+        }
+      );
+      websocket
+        .open()
+        .then(() => {
+          this.websocket = websocket;
+          resolve(this.websocket);
+        })
+        .catch((error) => {
+          console.error('failed to open websocket', error);
+          reject(
+            new Error(
+              'Failed to connect to API, confirm your server is online and try again.'
+            )
+          );
+        });
+    });
+
+    return this.connectPromise;
+  };
+
+  shutdown = async (): Promise<boolean> => {
+    if (this.websocket) {
+      const evt: CloseEvent = await this.websocket.close();
+      this.websocket = null;
+      return evt.type === 'close' && evt.wasClean;
+    }
+
+    return true;
+  };
+
+  getPassword = (): string | null => {
+    return sessionStorage.getItem(SESSION_STORAGE_KEY);
+  };
+
+  testPassword = async (password: string): Promise<boolean> => {
+    // Replace with password to check.
+    sessionStorage.setItem(SESSION_STORAGE_KEY, password);
+
+    // Attempt a "status" rpc call with the temporary password.
+    try {
+      await this.status();
+      return true;
+    } catch (err) {
+      // TODO: make sure error is auth error, not unrelated
+      this.clearPassword();
+      return false;
+    }
+  };
+
+  /*** Shared RPC methods */
+
+  status = async (): Promise<ServerStatus> => {
+    const statusResponse: StatusResponse = await this.rpc('status');
+    return statusResponse.server;
+  };
+
+  /*** Setup RPC methods ***/
+
+  setPassword = async (password: string): Promise<void> => {
+    sessionStorage.setItem(SESSION_STORAGE_KEY, password);
+
+    return this.rpc('set_password');
+  };
+
+  private clearPassword = () => {
+    sessionStorage.removeItem(SESSION_STORAGE_KEY);
+  };
+
+  setConfigGenConnections = async (
+    ourName: string,
+    leaderUrl?: string
+  ): Promise<void> => {
+    const connections = {
+      our_name: ourName,
+      leader_api_url: leaderUrl,
+    };
+
+    return this.rpc('set_config_gen_connections', connections);
+  };
+
+  getDefaultConfigGenParams = (): Promise<ConfigGenParams> => {
+    return this.rpc('get_default_config_gen_params');
+  };
+
+  getConsensusConfigGenParams = (): Promise<ConsensusState> => {
+    return this.rpc('get_consensus_config_gen_params');
+  };
+
+  setConfigGenParams = (params: ConfigGenParams): Promise<void> => {
+    return this.rpc('set_config_gen_params', params);
+  };
+
+  getVerifyConfigHash = (): Promise<PeerHashMap> => {
+    return this.rpc('get_verify_config_hash');
+  };
+
+  runDkg = (): Promise<void> => {
+    return this.rpc('run_dkg');
+  };
+
+  startConsensus = (): Promise<void> => {
+    // Special case: start_consensus kills the server. Set a timeout and restart after short period.
```

Comment:
Yeah so the tricky thing is that once you start consensus, the config server shuts down and the regular api starts up, so it necessarily kills the socket. Ideally the config server would stay up long enough to respond to the rpc, indicating that it successfully started the server, and then shuts down. Worth an issue to improve the understandability of the code IMO.

I haven't spent enough time here to know if this is a reasonable timeout, but you are correct that it could hypothetically not be enough time. Unfortunately while I was working on this, this part of starting consensus would cause the backend to panic and crash. Might be worth some retries if we find it's prone to taking longer than this.